File: php/JavaSlimeFinder.php
<?php

class JavaSlimeFinder {
    private const MULTIPLIER = 25214903917;
    private const MASK_48 = 281474976710655;
    private const ADDEND = 11;

    private static function toInt32($value) {
        return self::asIntN(32, $value);
    }

    private static function asIntN(int $bits, $value) {
        $val = gmp_init((string)$value);
        $mask = gmp_sub(gmp_pow(2, $bits), 1);
        $val = gmp_and($val, $mask);
        
        $threshold = gmp_pow(2, $bits - 1);
        if (gmp_cmp($val, $threshold) >= 0) {
            $val = gmp_sub($val, gmp_mul(2, $threshold));
        }
        return $val;
    }

    private static function next($seed, int $bits) {
        $multiplier = gmp_init(self::MULTIPLIER);
        $addend = gmp_init(self::ADDEND);
        $mask48 = gmp_init(self::MASK_48);
        
        $seed = gmp_and(gmp_add(gmp_mul($seed, $multiplier), $addend), $mask48);
        $shift = gmp_pow(2, 48 - $bits);
        return [$seed, gmp_div_q($seed, $shift)];
    }

    private static function nextInt($seed, $bound) {
        if ($bound <= 0) {
            throw new Exception("bound must be positive");
        }

        $boundBig = gmp_init((string)$bound);

        if (gmp_cmp(gmp_and($boundBig, gmp_sub($boundBig, 1)), 0) === 0) {
            $res = self::next($seed, 31);
            $mul = gmp_mul($boundBig, $res[1]);
            $div = gmp_div_q($mul, gmp_pow(2, 31));
            return [$res[0], $div];
        }

        $u = 0;
        $v = 0;
        $currentSeed = $seed;
        do {
            $res = self::next($currentSeed, 31);
            $currentSeed = $res[0];
            $u = $res[1];
            $v = gmp_mod($u, $boundBig);
        } while (gmp_cmp(gmp_add(gmp_sub($u, $v), gmp_sub($boundBig, 1)), 0) < 0);

        return [$currentSeed, $v];
    }

    public static function getSlimeRandomSeed($worldSeed, $x, $z) {
        $ws = self::asIntN(64, $worldSeed);
        $cx = $x;
        $cz = $z;

        $term1 = self::toInt32($cx * $cx * 0x4c1906);
        $term2 = self::toInt32($cx * 0x5ac0db);
        $term3 = self::toInt32($cz * $cz) * 0x4307a7;
        $term4 = self::toInt32($cz * 0x5f24f);
        $xorVal = 0x3ad8025f;

        $seed = $ws + $term1 + $term2 + $term3 + $term4;
        $seed = $seed ^ $xorVal;

        return $seed;
    }

    public static function isSlimeChunk($worldSeed, $x, $z) {
        $seed = self::getSlimeRandomSeed($worldSeed, $x, $z);
        $seed = self::asIntN(64, $seed);

        $multiplier = gmp_init(self::MULTIPLIER);
        $currentSeed = gmp_and(gmp_xor($seed, $multiplier), gmp_init(self::MASK_48));
        $currentSeed = self::asIntN(48, $currentSeed);

        $res = self::nextInt($currentSeed, 10);

        return gmp_cmp($res[1], 0) === 0;
    }
}

File: python/java_slime_finder.py
class JavaSlimeFinder:
    _MULTIPLIER = 0x5DEECE66D
    _MASK_48 = (1 << 48) - 1
    _ADDEND = 0xB

    @staticmethod
    def _to_int32(value: int) -> int:
        value &= 0xFFFFFFFF
        if value & 0x80000000:
            return value - 0x100000000
        return value

    @staticmethod
    def _next(seed: int, bits: int) -> tuple[int, int]:
        seed = ((seed * JavaSlimeFinder._MULTIPLIER) + JavaSlimeFinder._ADDEND) & JavaSlimeFinder._MASK_48
        return seed, seed >> (48 - bits)

    @staticmethod
    def _next_int(seed: int, bound: int) -> tuple[int, int]:
        if bound <= 0:
            raise ValueError("bound must be positive")

        if (bound & (bound - 1)) == 0:
            seed, res = JavaSlimeFinder._next(seed, 31)
            return seed, (bound * res) >> 31

        while True:
            seed, u = JavaSlimeFinder._next(seed, 31)
            v = u % bound
            if u - v + (bound - 1) >= 0:
                return seed, v

    @classmethod
    def get_slime_random_seed(cls, world_seed: int, x: int, z: int) -> int:
        ws = world_seed & 0xFFFFFFFFFFFFFFFF
        if ws & 0x8000000000000000:
            ws -= 0x10000000000000000

        cx = x
        cz = z

        term1 = cls._to_int32(cx * cx * 0x4C1906)
        term2 = cls._to_int32(cx * 0x5AC0DB)
        term3 = cls._to_int32(cz * cz) * 0x4307A7
        term4 = cls._to_int32(cz * 0x5F24F)
        xor_val = 0x3AD8025F

        seed = ws + term1 + term2 + term3 + term4
        seed ^= xor_val

        return seed & 0xFFFFFFFFFFFFFFFF

    @classmethod
    def is_slime_chunk(cls, world_seed: int, x: int, z: int) -> bool:
        seed = cls.get_slime_random_seed(world_seed, x, z)

        if seed & 0x8000000000000000:
            seed -= 0x10000000000000000

        current_seed = (seed ^ cls._MULTIPLIER) & cls._MASK_48

        _, result = cls._next_int(current_seed, 10)

        return result == 0